source/gui/contextHelp.py

```diff
@@ -0,0 +1,88 @@
+# -*- coding: UTF-8 -*-
+# A part of NonVisual Desktop Access (NVDA)
+# Copyright (C) 2017-2020 NV Access Limited, Thomas Stivers
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+import os
+import tempfile
+import typing
+
+import gui
+import ui
+import wx
+from logHandler import log
+
+
+def writeRedirect(helpId: str, helpFilePath: str, contextHelpPath: str):
+	redirect = rf"""
+<html><head>
+<meta http-equiv="refresh" content="0;url=file:///{helpFilePath}#{helpId}" />
+</head></html>
+	"""
+	with open(contextHelpPath, 'w') as f:
+		f.write(redirect)
+
+
+def showHelp(helpId: str):
+	"""Display the corresponding section of the user guide when either the Help
+	button in an NVDA dialog is pressed or the F1 key is pressed on a
+	recognized control.
+	"""
+	if not helpId:
+		# Translators: Message indicating no context sensitive help is available.
+		noHelpMessage = _("No context sensitive help is available here at this time.")
+		ui.message(noHelpMessage)
+	helpFile = gui.getDocFilePath("userGuide.html")
+	if not os.path.exists(helpFile):
+		# Translators: Message shown when trying to display context sensitive help,
+		# indicating that	the user guide could not be found.
+		noHelpMessage = _("No user guide found.")
+		log.debugWarning("No user guide found: possible cause - running from source without building user docs")
+		ui.message(noHelpMessage)
+	log.debug(f"Opening help: helpId = {helpId}, userGuidePath: {helpFile}")
+
+	nvdaTempDir = os.path.join(tempfile.gettempdir(), "nvda")
+	if not os.path.exists(nvdaTempDir):
+		os.mkdir(nvdaTempDir)
+
+	contextHelpRedirect = os.path.join(nvdaTempDir, "contextHelp.html")
+	try:
+		writeRedirect(helpId, helpFile, contextHelpRedirect)
+	except Exception:
+		log.error("Unable to write context help redirect file.", exc_info=True)
+		return
+
+	try:
+		os.startfile(f"file://{contextHelpRedirect}")
+	except Exception:
+		log.error("Unable to launch context help.", exc_info=True)
+
+
+def bindHelpEvent(helpId: str, window: wx.Window):
+	window.Unbind(wx.EVT_HELP)
+	window.Bind(
+		wx.EVT_HELP,
+		lambda evt: _onEvtHelp(helpId, evt),
+	)
+	log.debug(f"Did context help binding for {window.__class__.__qualname__}")
+
+
+def _onEvtHelp(helpId: str, evt: wx.HelpEvent):
+	# Don't call evt.skip. Events bubble upwards through parent controls.
+	# Context help for more specific controls should override the less specific parent controls.
+	showHelp(helpId)
+
+
+class ContextHelpMixin:
+	def __init__(self, *args, **kwargs):
+		super().__init__(*args, **kwargs)
+		helpId = getattr(self, "helpId", None)

```

Comment:
I think helpId could be implemented as an abstract property. I also think every panel needs the mixin, because add-ons can't provide context help.
In any case, we should avoid that something that doesn't have help has the mixin.

Reply:
I'm sorry, I don't follow. I have already added the mixin to the settingsPanel base class.

Reply:
Ah, I of course should have been clearer.

### Current situation, mixin added to SettingsPanel
* Every SettingsPanel has the mixin, so is expected to provide context help
* This doesn't apply to add-ons that use SettingsPanel, and there are plenty of them nowadays.

### SUggested situation
* Remove mixin from SettingsPanel
* Add mixin to every SettingsPanel NVDA core offers. Could be done using a SettingsPanelWithContextHelp intermediate class that inherrits from SettingsPanel and the mixin, though I don't care about being a bit more explicit here
* Then, mark helpId as an abstract property. It ensures that every panel or dialog that doesn't have helpId simply fails to initialize.

```
	@abstractproperty
	def helpId(self):
		raise NotImplementedError

```

Reply:
I'd prefer not to do that. I think it's fair that every settings panel should have help, and for users to be informed that the addon panels do not have context help available.

Reply:
Currently, users are only informed with a warning in the log. I'd agree if the help event would be bound regardless and show a message box indicating that there's no help.
Otherwise, adding a mixing to something that does nothing with it honestly really feels a bit strange to me.

Reply:
> Currently, users are only informed with a warning in the log

Ah yes, this is a mistake, note that there is a user message when the helpId is empty.